### ISSUE TITLE
FP S6507 when using Parallel.ForEach

### ISSUE THREAD

**Issue by paulhickman-a365** [SonarSource/sonar-dotnet#7058]
The following code raises S6507 "Blocks should not be synchronized on local variables" on the line ````lock (l)````

````c#
public int Total(int[] values)
{
    object l = new object();
    int total = 0;

    Parallel.ForEach(values, v => {
        lock (l)
        { 
            total += v; 
        }
    });

    return total;
}
````

However, when:

- the lock statement is in a lambda
- the local variable is in captured scope
- the code in scope of the lock does not access any members of the parent class
- the lambda is an argument to a framework method known to introduce parallelism. 

using a local variable probably gives the behavior the developer intends.

In this example, locking on a member variable of the class would cause unnecessary blocking between calls to the method.

Language C#, .Net 6
Sonar Analzyer Version: 8.55.0.65544
Visual Studio: 17.5.3

**Comment by cristian-ambrosini-sonarsource** [comment]
Thank you for opening this issue, @paulhickman-a365. Your feedback helps us improve our products. We will look at it as soon as possible. After we confirm the issue, we will add labels and put it in our backlog

**Comment by antonioaversa** [comment]
@paulhickman-a365 Thank you for reporting the issue. I confirm it as FP. 
We plan to not trigger the rule when the local variable being locked is in the captured scope of a lambda.
That should address the problem reported.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1832](https://sonarsource.atlassian.net/browse/NET-1832)

[NET-1832]: https://sonarsource.atlassian.net/browse/NET-1832?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
